File: simplejavable/java/src/main/java/org/simplejavable/Adapter.java
package org.simplejavable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

/**
 * Represents a physical Bluetooth adapter on the host system.
 */
public class Adapter {
    private EventListener eventListener;
    private final long instanceId;

    private final Callback callbacks = new Callback() {
        @Override
        public void onScanStart() {
            if (eventListener != null) {
                eventListener.onScanStart();
            }
        }

        @Override
        public void onScanStop() {
            if (eventListener != null) {
                eventListener.onScanStop();
            }
        }

        @Override
        public void onScanUpdated(long peripheralId) {
            if (eventListener != null) {
                eventListener.onScanUpdated(new Peripheral(instanceId, peripheralId));
            }
        }

        @Override
        public void onScanFound(long peripheralId) {
            if (eventListener != null) {
                eventListener.onScanFound(new Peripheral(instanceId, peripheralId));
            }
        }
    };

    private Adapter(long newInstanceId) {
        this.instanceId = newInstanceId;
        nativeAdapterRegister(instanceId, callbacks);
    }

    /**
     * Gets the system identifier of the Bluetooth adapter.
     * @return The string identifier.
     */
    public String getIdentifier() {
        String identifier = nativeAdapterIdentifier(instanceId);
        return identifier != null ? identifier : "";
    }

    /**
     * Gets the MAC address of the Bluetooth adapter.
     * @return The BluetoothAddress.
     */
    public BluetoothAddress getAddress() {
        String address = nativeAdapterAddress(instanceId);
        return new BluetoothAddress(address != null ? address : "");
    }

    /**
     * Starts scanning for BLE peripherals asynchronously.
     */
    public void scanStart() {
        nativeAdapterScanStart(instanceId);
    }

    /**
     * Stops an active BLE peripheral scan.
     */
    public void scanStop() {
        nativeAdapterScanStop(instanceId);
    }

    /**
     * Starts scanning for BLE peripherals, blocking for the specified duration.
     * @param timeoutMs The scan duration in milliseconds.
     * @throws Exception If the scan fails.
     */
    public void scanFor(int timeoutMs) throws Exception {
        nativeAdapterScanFor(instanceId, timeoutMs);
    }

    /**
     * Starts scanning for BLE peripherals asynchronously and returns the results when finished.
     * @param timeoutMs The scan duration in milliseconds.
     * @return A CompletableFuture containing the list of found peripherals.
     */
    public CompletableFuture<List<Peripheral>> scanForAsync(int timeoutMs) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                scanFor(timeoutMs);
                return scanGetResults();
            } catch (Exception e) {
                throw new CompletionException(e);
            }
        });
    }

    /**
     * Checks if the adapter is currently scanning.
     * @return True if a scan is active.
     */
    public boolean getScanIsActive() {
        return nativeAdapterScanIsActive(instanceId);
    }

    /**
     * Gets the list of peripherals found during the last scan.
     * @return A list of Peripheral objects.
     */
    public List<Peripheral> scanGetResults() {
        long[] results = nativeAdapterScanGetResults(instanceId);
        List<Peripheral> peripherals = new ArrayList<>();
        for (long id : results) {
            peripherals.add(new Peripheral(instanceId, id));
        }
        return peripherals;
    }

    /**
     * Sets the event listener for adapter events (e.g., scan start, stop, found).
     * @param listener The EventListener implementation.
     */
    public void setEventListener(EventListener listener) {
        this.eventListener = listener;
    }

    /**
     * Gets a list of all peripherals currently paired with the system.
     * @return A list of paired Peripheral objects.
     */
    public List<Peripheral> getPairedPeripherals() {
        long[] results = nativeAdapterGetPairedPeripherals(instanceId);
        List<Peripheral> peripherals = new ArrayList<>();
        for (long id : results) {
            peripherals.add(new Peripheral(instanceId, id));
        }
        return peripherals;
    }

    /**
     * Checks if Bluetooth is enabled on the host system.
     * @return True if Bluetooth is enabled.
     */
    public static boolean isBluetoothEnabled() {
        return nativeIsBluetoothEnabled();
    }

    /**
     * Gets a list of all available Bluetooth adapters on the system.
     * @return A list of Adapter objects.
     */
    public static List<Adapter> getAdapters() {
        long[] nativeAdapterIds = nativeGetAdapters();
        List<Adapter> adapters = new ArrayList<>();

        for (long nativeAdapterId : nativeAdapterIds) {
            adapters.add(new Adapter(nativeAdapterId));
        }

        return adapters;
    }

    // Native method declarations
    private static native long[] nativeGetAdapters();
    private static native boolean nativeIsBluetoothEnabled();
    private native void nativeAdapterRegister(long adapterId, Callback callback);
    private native String nativeAdapterIdentifier(long adapterId);
    private native String nativeAdapterAddress(long adapterId);
    private native void nativeAdapterScanStart(long adapterId);
    private native void nativeAdapterScanStop(long adapterId);
    private native void nativeAdapterScanFor(long adapterId, int timeout);
    private native boolean nativeAdapterScanIsActive(long adapterId);
    private native long[] nativeAdapterScanGetResults(long adapterId);
    private native long[] nativeAdapterGetPairedPeripherals(long adapterId);

    private interface Callback {
        void onScanStart();
        void onScanStop();
        void onScanUpdated(long peripheralId);
        void onScanFound(long peripheralId);
    }

    public interface EventListener {
        default void onScanStart() {}
        default void onScanStop() {}
        default void onScanUpdated(Peripheral peripheral) {}
        default void onScanFound(Peripheral peripheral) {}
    }

    static {
        try {
            NativeLibraryLoader.loadLibrary("simplejavable");
        } catch (IOException e) {
            throw new RuntimeException("Failed to load native library", e);
        }

        // Runtime.getRuntime().addShutdownHook(new Thread(() -> {
        //     // TODO: We might need this to clean up the native library.
        //     // System.out.println("JVM shutdown initiated at " + System.currentTimeMillis());
        // }));
    }
}

File: simpledroidble/simpledroidble/src/main/java/org/simpleble/android/Adapter.kt
package org.simpleble.android

import android.content.Context
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class Adapter private constructor(private val instanceId: Long) {
    private val _onScanStart = MutableSharedFlow<Unit>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    private val _onScanStop = MutableSharedFlow<Unit>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    private val _onScanActive = MutableSharedFlow<Boolean>(replay = 1, extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    private val _onScanUpdated = MutableSharedFlow<Peripheral>(extraBufferCapacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)
    private val _onScanFound = MutableSharedFlow<Peripheral>(extraBufferCapacity = 64, onBufferOverflow = BufferOverflow.DROP_OLDEST)

    private val peripherals = ConcurrentHashMap<Long, Peripheral>()

    private fun peripheral(peripheralId: Long): Peripheral {
        return peripherals.computeIfAbsent(peripheralId) { Peripheral(instanceId, it) }
    }

    private val callbacks = object : Callback {
        override fun onScanStart() {
            _onScanStart.tryEmit(Unit)
            _onScanActive.tryEmit(true)
        }

        override fun onScanStop() {
            _onScanStop.tryEmit(Unit)
            _onScanActive.tryEmit(false)
        }

        override fun onScanUpdated(peripheralId: Long) {
            _onScanUpdated.tryEmit(peripheral(peripheralId))
        }

        override fun onScanFound(peripheralId: Long) {
            _onScanFound.tryEmit(peripheral(peripheralId))
        }
    }

    init { nativeAdapterRegister(instanceId, callbacks) }

    val identifier: String get() = nativeAdapterIdentifier(instanceId)

    val address: BluetoothAddress get() = BluetoothAddress(nativeAdapterAddress(instanceId))

    fun scanStart() {
        nativeAdapterScanStart(instanceId)
    }

    fun scanStop() {
        nativeAdapterScanStop(instanceId)
    }

    suspend fun scanFor(timeoutMs: Int) {
        withContext(Dispatchers.IO) {
            nativeAdapterScanFor(instanceId, timeoutMs)
        }
    }

    val scanIsActive: Boolean get() {
        return nativeAdapterScanIsActive(instanceId)
    }

    fun scanGetResults(): List<Peripheral> {
        return nativeAdapterScanGetResults(instanceId).map(::peripheral)
    }

    val onScanStart: SharedFlow<Unit> get() = _onScanStart

    val onScanStop: SharedFlow<Unit> get() = _onScanStop

    val onScanActive: SharedFlow<Boolean> get() = _onScanActive

    val onScanUpdated: SharedFlow<Peripheral> get() = _onScanUpdated

    val onScanFound: SharedFlow<Peripheral> get() = _onScanFound

    fun getPairedPeripherals(): List<Peripheral> {
        return nativeAdapterGetPairedPeripherals(instanceId).map(::peripheral)
    }

    /**
     * Create a BLE peripheral hosted by this Android application.
     *
     * The application context is retained by the native backend; the supplied
     * Activity is not retained. Configure services and advertising data on the
     * returned object before starting it.
     */
    fun createLocalPeripheral(context: Context): LocalPeripheral {
        return LocalPeripheral(nativeAdapterCreateLocalPeripheral(instanceId, context.applicationContext))
    }

    companion object {
        init {
            SimpleDroidBle.ensureLoaded()
        }

        @JvmStatic
        fun isBluetoothEnabled(): Boolean {
            return nativeIsBluetoothEnabled()
        }

        @JvmStatic
        fun getAdapters(): List<Adapter> {
            return nativeGetAdapters().map { adapterId ->
                adapterCache.computeIfAbsent(adapterId, ::Adapter)
            }
        }

        private val adapterCache = ConcurrentHashMap<Long, Adapter>()

        @JvmStatic
        private external fun nativeGetAdapters(): LongArray

        private external fun nativeIsBluetoothEnabled(): Boolean
    }

    // ----------------------------------------------------------------------------

    private external fun nativeAdapterRegister(adapterId: Long, callback: Callback)

    private external fun nativeAdapterIdentifier(adapterId: Long): String

    private external fun nativeAdapterAddress(adapterId: Long): String

    private external fun nativeAdapterScanStart(adapterId: Long)

    private external fun nativeAdapterScanStop(adapterId: Long)

    private external fun nativeAdapterScanFor(adapterId: Long, timeout: Int)

    private external fun nativeAdapterScanIsActive(adapterId: Long): Boolean

    private external fun nativeAdapterScanGetResults(adapterId: Long) : LongArray

    private external fun nativeAdapterGetPairedPeripherals(adapterId: Long): LongArray

    private external fun nativeAdapterCreateLocalPeripheral(adapterId: Long, applicationContext: Context): Long

    // ----------------------------------------------------------------------------

    private interface Callback {
        fun onScanStart()
        fun onScanStop()
        fun onScanUpdated(peripheralId: Long)
        fun onScanFound(peripheralId: Long)
    }

}
